src/realm/transaction.cpp

```diff
@@ -45,32 +44,71 @@ ColInfo get_col_info(const Table* table)
     return cols;
 }
 
+void add_list_to_repl(CollectionBase& list, Replication& repl, util::UniqueFunction<void(Mixed)> update_embedded);
+void add_dictionary_to_repl(Dictionary& dict, Replication& repl, util::UniqueFunction<void(Mixed)> update_embedded)
+{
+    size_t sz = dict.size();
+    for (size_t n = 0; n < sz; ++n) {
+        const auto& [key, val] = dict.get_pair(n);
+        repl.dictionary_insert(dict, n, key, val);
+        if (val.is_type(type_List)) {
+            auto n_list = dict.get_list({key.get_string()});
+            add_list_to_repl(*n_list, *dict.get_table()->get_repl(), std::move(update_embedded));
+        }
+        else if (val.is_type(type_Dictionary)) {
+            repl.dictionary_insert(dict, n, key, val);
+            auto n_dict = dict.get_dictionary({key.get_string()});
+            add_dictionary_to_repl(*n_dict, *dict.get_table()->get_repl(), std::move(update_embedded));
+        }
+        else if (update_embedded) {
+            update_embedded(val);
+        }
+    }
+}
+
+void add_list_to_repl(CollectionBase& list, Replication& repl, util::UniqueFunction<void(Mixed)> update_embedded)
+{
+    auto sz = list.size();
+    for (size_t n = 0; n < sz; n++) {
+        auto val = list.get_any(n);
+        repl.list_insert(list, n, val, n);
+        if (val.is_type(type_List)) {
+            auto n_list = list.get_list({n});
+            add_list_to_repl(*n_list, *list.get_table()->get_repl(), std::move(update_embedded));
+        }
+        else if (val.is_type(type_Dictionary)) {
+            auto n_dict = list.get_dictionary({n});
+            add_dictionary_to_repl(*n_dict, *list.get_table()->get_repl(), std::move(update_embedded));
+        }
+        else if (update_embedded) {
+            update_embedded(val);
+        }
+    }
+}
+
 void generate_properties_for_obj(Replication& repl, const Obj& obj, const ColInfo& cols)
 {
     for (auto elem : cols) {
         auto col = elem.first;
         auto embedded_table = elem.second;
-        auto cols_2 = get_col_info(embedded_table);
-        auto update_embedded = [&](Mixed val) {
-            if (val.is_null()) {
+        util::UniqueFunction<void(Mixed)> update_embedded = nullptr;
+        if (embedded_table) {
+            update_embedded = [&](Mixed val) {
+                if (val.is_null()) {
+                    return;
+                }
+                REALM_ASSERT(val.is_type(type_Link, type_TypedLink));
+                Obj embedded_obj = embedded_table->get_object(val.get<ObjKey>());
+                auto cols_2 = get_col_info(embedded_table);
```

Comment:
You should not generate this info for every call. This should be done before the function is created ahd thus captured. So you have to declare cols_2 before the if (embedded_table), Refer to my initial comment : https://github.com/realm/realm-core/pull/6714#discussion_r1231025544